src/transformers/models/superpoint/modeling_superpoint.py

```diff
@@ -0,0 +1,577 @@
+        input = self.pool(input)
+
+        if output_hidden_states:
+            all_hidden_states = all_hidden_states + (input,)
+
+        input = self.relu(self.conv4a(input))
+        output = self.relu(self.conv4b(input))
+
+        if output_hidden_states:
+            all_hidden_states = all_hidden_states + (output,)
+
+        if not return_dict:
+            return tuple(v for v in [output, all_hidden_states] if v is not None)
+
+        return BaseModelOutputWithNoAttention(
+            last_hidden_state=output,
+            hidden_states=all_hidden_states,
+        )
+
+
+class SuperPointInterestPointDecoder(nn.Module):
+    """
+    The SuperPointInterestPointDecoder uses the output of the SuperPointEncoder to compute the keypoint with scores.
+    The scores are first computed by a convolutional layer, then a softmax is applied to get a probability distribution
+    over the 65 possible keypoint classes. The keypoints are then extracted from the scores by thresholding and
+    non-maximum suppression. Post-processing is then applied to remove keypoints too close to the image borders as well
+    as to keep only the k keypoints with highest score.
+    """
+
+    def __init__(self, config: SuperPointConfig) -> None:
+        super().__init__()
+        self.hidden_sizes = config.hidden_sizes
+        self.descriptor_dim = config.descriptor_dim
+        self.keypoint_threshold = config.keypoint_threshold
+        self.max_keypoints = config.max_keypoints
+        self.nms_radius = config.nms_radius
+        self.border_removal_distance = config.border_removal_distance
+
+        self.relu = nn.ReLU(inplace=True)
+        self.pool = nn.MaxPool2d(kernel_size=2, stride=2)
+        self.convSa = nn.Conv2d(
+            self.hidden_sizes[3],
+            self.hidden_sizes[4],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+        self.convSb = nn.Conv2d(self.hidden_sizes[4], 65, kernel_size=1, stride=1, padding=0)
+
+    def forward(self, encoded: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        # Compute the dense keypoint scores
+        scores = self.get_scores(encoded)
+        # Extract keypoints
+        keypoints, scores = self.extract_keypoints(scores)
+
+        return keypoints, scores
+
+    def get_scores(self, encoded: torch.Tensor) -> torch.Tensor:
+        """Compute the dense keypoint scores"""
+        scores = self.relu(self.convSa(encoded))
+        scores = self.convSb(scores)
+        scores = torch.nn.functional.softmax(scores, 1)[:, :-1]
+        batch_size, _, height, width = scores.shape
+        scores = scores.permute(0, 2, 3, 1).reshape(batch_size, height, width, 8, 8)
+        scores = scores.permute(0, 1, 3, 2, 4).reshape(batch_size, height * 8, width * 8)
+        scores = self.simple_nms(scores, self.nms_radius)
+        return scores
+
+    def extract_keypoints(self, scores: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
```

Comment:
Lets make these private methods: we want to keep the public interface of our models' as consistent as possible

Reply:
Fixed in [this commit](https://github.com/huggingface/transformers/pull/25786/commits/53c33488a366ad4da43042fef273b70a8aee413b)